spark/src/main/scala/geotrellis/spark/summary/polygonal/RDDPolygonalSummary.scala

```diff
@@ -0,0 +1,143 @@
+ * limitations under the License.
+ */
+
+package geotrellis.spark.summary.polygonal
+
+import java.util.UUID
+
+import cats.Semigroup
+import cats.syntax.semigroup._
+import geotrellis.raster._
+import geotrellis.raster.rasterize.Rasterizer
+import geotrellis.raster.summary.GridVisitor
+import geotrellis.raster.summary.polygonal._
+import geotrellis.layer.{Metadata, SpatialKey, TileLayerMetadata}
+import geotrellis.util.MethodExtensions
+import geotrellis.vector._
+import org.apache.spark.SparkContext
+import org.apache.spark.rdd.RDD
+
+import scala.reflect.ClassTag
+
+object RDDPolygonalSummary {
+  /**
+    * Perform a polygonal summary over all geometries in geometryRdd using data from rasterRdd
+    *
+    * Helper methods are provided in [[RDDPolygonalSummaryMethods]] which handle some common
+    * derived cases, such as only performing a summary for a single geometry across rasterRdd.
+    *
+    * The provided visitor is cloned for each individual polygonal summary operation on each
+    * geometry in the input. Do not rely on private state within the passed visitor to persist
+    * to each summary operation.
+    *
+    * @param rasterRdd
+    * @param geometryRdd
+    * @param visitor
+    * @param options Rasterizer.Options for this polygonal summary operation
+    * @param semigroupR
+    * @param tagR
+    * @tparam R Type of the summary being performed
+    * @tparam T Type of the Raster contained in rasterRdd
+    * @return An RDD containing each unique input geometry and it's summary result R
+    */
+  def apply[R, T <: CellGrid[Int]](
+      rasterRdd: RDD[(SpatialKey, T)] with Metadata[TileLayerMetadata[SpatialKey]],
+      geometryRdd: RDD[Geometry],
+      visitor: GridVisitor[Raster[T], R],
+      options: Rasterizer.Options
+  )(implicit semigroupR: Semigroup[R],
+    tagR: ClassTag[R]): RDD[Feature[Geometry, PolygonalSummaryResult[R]]] = {
+    val layout = rasterRdd.metadata.layout
+
+    val keyedGeometryRdd: RDD[(SpatialKey, Feature[Geometry, UUID])] =
+      geometryRdd.flatMap { geom =>
+        val uuid = UUID.randomUUID
+        val keys: Set[SpatialKey] = layout.mapTransform.keysForGeometry(geom)
+        keys.map(k => (k, Feature(geom, uuid)))
+      }
+
+    // Combine geometries and rasters by SpatialKey, then loop partitions
+    //  and perform the polygonal summary on each combination
+    val joinedRdd = keyedGeometryRdd.join(rasterRdd)
+    val featuresWithSummaries: RDD[(UUID, Feature[Geometry, PolygonalSummaryResult[R]])] =
+      joinedRdd.map { data => {
+        val spatialKey = data._1
+        val feature: Feature[Geometry, UUID] = data._2._1
+        val tile: T = data._2._2
+        val extent: Extent = layout.mapTransform.keyToExtent(spatialKey)
+        val raster: Raster[T] = Raster(tile, extent)
+        val result: PolygonalSummaryResult[R] =
+          raster.polygonalSummary(feature.geom, visitor.getClass.newInstance, options)
```

Comment:
So this is getting a new instance of the visitor class, not of type R. I'm not sure how else to specify the visitor class you want to use and then also be able to construct new instances of it internally. Perhaps we could add a `def reset()` method to GridVisitor that should be implemented so that we can always use the passed instance? That feels pretty icky though because then we're getting farther into the weeds of mutating the internal state of a thing in a distributed environment. 

> we can use Monoid[R].empty to get a new instance of visitor

It's unclear to me how that would work. Could you perhaps explain what you're thinking in a different way?

Reply:
Right .. you're totally right. Thats what I get while reviewing PRs and doing something else.

I believe this was in response to the problem of having a visitor with some initial state being passed into the the summary and skewing the results in a surprising way?

Reply:
I don't know that there is a great way out of this. 

Adding `def map(v: R): R` to `GridVisitor` might be one way, that way you can get back to use the `Monoid[R].empty`. Adding `Functor[GridVisitor[Raster[T], R]` would be another, more abstract way to add that constraint.

Actually all of the `R` instances already have `Monoid[R]` so you're only a step away from `Monoid[GridVisitor[Raster[T], R]` .. which is super weird because its a functional composition over dirty mutable state. This may have pay-off that some point that is not now.

I guess the thing to do is document the requirement for zero-argument constructor in `GridVisitor` doctoring and see if it ever comes up again.

Reply:
> this was in response to the problem of having a visitor with some initial state being passed into the the summary

Yup. The visitor internal state wasn't reset on each iteration of the polygonal summary call, which we make a number of times for each combination of spatial key, raster and geometry. 

> I guess the thing to do is document the requirement for zero-argument constructor in GridVisitor doc-string and see if it ever comes up again.

Seems like the best out ATM. I'm on board with that. I like that better than some workaround that leads to weird function signatures and more complicated implementation requirements for an optimization we may never need.